_posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd

```diff
@@ -0,0 +1,740 @@
+one). These devices then send model outputs to a central server, where final results are computed (stage two). Sure, you will
+be thinking, this is a convenient setup for our scenario: If we intercept stage one results, we -- most probably -- gain
+access to richer information than what is contained in a model's final output layer. --- That is correct, but the scenario is
+less contrived than one might assume. Just like federated learning [@McMahanMRA16], it fulfills important desiderata: Actual
+training data never leaves the devices, thus staying (in theory!) private; at the same time, ingoing traffic to the server is
+significantly reduced.
+
+In our example setup, the on-device model is a convnet, while the server model is a simple feedforward network.
+
+We link both together as a *TargetModel* that when called normally, will run both steps in succession. However, we'll be able
+to call `target_model$mobile_step()` separately, thus intercepting intermediate results.
+
+```{r}
+on_device_model <- keras_model_sequential() %>%
+  layer_conv_2d(filters = 32, kernel_size = c(7, 7),
+                input_shape = c(105, 105, 1), activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_max_pooling_2d(pool_size = c(3, 3), strides = 3) %>%
+  layer_dropout(0.2) %>%
+  layer_conv_2d(filters = 32, kernel_size = c(7, 7), activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_max_pooling_2d(pool_size = c(3, 3), strides = 2) %>%
+  layer_dropout(0.2) %>%
+  layer_conv_2d(filters = 32, kernel_size = c(5, 5), activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_max_pooling_2d(pool_size = c(2, 2), strides = 2) %>%
+  layer_dropout(0.2) %>%
+  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_max_pooling_2d(pool_size = c(2, 2), strides = 2) %>%
+  layer_dropout(0.2) 
+
+server_model <- keras_model_sequential() %>%
+  layer_dense(units = 256, activation = "relu") %>%
+  layer_flatten() %>%
+  layer_dropout(0.2) %>% 
+  # we have just 20 different ids, but they are not in lexicographic order
+  layer_dense(units = 50, activation = "softmax")
+
+target_model <- function() {
+  keras_model_custom(name = "TargetModel", function(self) {
+    
+    self$on_device_model <-on_device_model
+    self$server_model <- server_model
+    self$mobile_step <- function(inputs) 
+      self$on_device_model(inputs)
+    self$server_step <- function(inputs)
+      self$server_model(inputs)
+
+    function(inputs, mask = NULL) {
+      inputs %>% 
+        self$mobile_step() %>%
+        self$server_step()
+    }
+  })
+  
+}
+
+model <- target_model()
+
+```
+
+The overall model is a Keras custom model, so we train it [TensorFlow 2.x -
+style](https://tensorflow.rstudio.com/tutorials/advanced/). After ten epochs, training and validation accuracy are at
+~0.84\ and~ 0.73, respectively -- not bad at all for a 20-class discrimination task.
```

Comment:
This seems to be causing some weird rendering, intended?

Reply:
right, thanks! Will correct...